packages/datadog-plugin-jest/src/index.js

```diff
@@ -0,0 +1,202 @@
+    constructor (config, context) {
+      super(config, context)
+      this.testSuite = context.testPath.replace(`${config.rootDir}/`, '')
+    }
+  }
+}
+
+function createWrapTeardown (tracer) {
+  return function wrapTeardown (teardown) {
+    return async function teardownWithTrace () {
+      await new Promise((resolve) => {
+        tracer._exporter._writer.flush(resolve)
+      })
+      return teardown.apply(this, arguments)
+    }
+  }
+}
+
+function createHandleTestEvent (tracer, testMetadata) {
+  return async function (event) {
+    if (event.name === 'test_skip' || event.name === 'test_todo') {
+      const childOf = tracer.extract('text_map', {
+        'x-datadog-trace-id': id().toString(10),
+        'x-datadog-parent-id': '0000000000000000',
+        'x-datadog-sampled': 1
+      })
+      tracer.startSpan(
+        'jest.test',
+        {
+          childOf,
+          tags: {
+            [SPAN_TYPE]: 'test',
+            [RESOURCE_NAME]: `${event.test.parent.name}.${event.test.name}`,
+            [TEST_TYPE]: 'test',
+            [TEST_NAME]: event.test.name,
+            [TEST_SUITE]: this.testSuite,
+            [TEST_STATUS]: 'skip',
+            [SAMPLING_RULE_DECISION]: 1,
+            ...testMetadata
+          }
+        }
+      ).finish()
+    }
+    if (event.name === 'test_start') {
+      const childOf = tracer.extract('text_map', {
+        'x-datadog-trace-id': id().toString(10),
+        'x-datadog-parent-id': '0000000000000000',
+        'x-datadog-sampled': 1
+      })
+      let specFunction = event.test.fn
+      if (specFunction.length) {
+        specFunction = promisify(specFunction)
+      }
+      event.test.fn = tracer.wrap(
+        'jest.test',
+        { type: 'test',
+          childOf,
+          resource: `${event.test.parent.name}.${event.test.name}`,
+          tags: {
+            [TEST_TYPE]: 'test',
+            [TEST_NAME]: event.test.name,
+            [TEST_SUITE]: this.testSuite,
+            [SAMPLING_RULE_DECISION]: 1,
+            ...testMetadata
+          } },
+        async () => {
+          let result
+          try {
+            result = await specFunction()
+            tracer.scope().active().setTag(TEST_STATUS, 'pass')
+          } catch (error) {
+            tracer.scope().active().setTag(TEST_STATUS, 'fail')
+            throw error
+          } finally {
+            tracer
+              .scope()
+              .active()
+              .context()._trace.started.forEach((span) => {
+                span.finish()
+              })
+          }
+          return result
+        }
+      )
+    }
+  }
+}
+
+module.exports = [
+  {
+    name: 'jest-environment-node',
+    versions: ['>=24.8.0'],
+    patch: function (NodeEnvironment, tracer) {
+      const testMetadata = getTestMetadata()
+
+      this.wrap(NodeEnvironment.prototype, 'teardown', createWrapTeardown(tracer))
+      NodeEnvironment.prototype.handleTestEvent = createHandleTestEvent(tracer, testMetadata)
```

Comment:
this is not by default on `NodeEnvironment.prototype`. jest checks if it exists in https://github.com/facebook/jest/blob/c8418dfab714ddd75dcaaca28f9811c7da3e6123/packages/jest-circus/src/legacy-code-todo-rewrite/jestAdapterInit.ts#L121. Therefore I can't wrap it (it fails). I do delete the method in the unpatch though

Reply:
Not sure I understand, how can it not be on the prototype if it's used? Is it added in the constructor?

Reply:
it is a bit weird. In the [docs](https://jestjs.io/docs/en/configuration#testenvironment-string) it says:

> The class may optionally expose an asynchronous handleTestEvent method to bind to events fired by jest-circus

so if you want to implement your own test environment, you may optionally implement this method. In our case, we wrap the default environments which happen to *not* include this method.

Reply:
to clarify (spoken in slack as well): I meant I can't use `this.wrap` with an `undefined` property because it fails